File: web/src/entities/bossWall.ts
import { Direction } from '../constants.js';
import { ImmutableVec2 } from '../math.js';
import { Entity } from '../entity.js';
import { Resource } from '../resource.js';
import type { Animation } from '../media/animation.js';
import type { Room } from '../room.js';
import { BossSaw } from './bossSaw.js';

export class BossWall extends Entity {
  private readonly wall: Animation;
  private readonly direction: Direction;
  private frameCounter = 0;
  private active = false;

  constructor(direction: Direction) {
    super();
    this.direction = direction;
    this.setSize(new ImmutableVec2(10, 100));
    this.wall = Resource.getAnimation('data/images/wall.bmp', 1);
  }

  getLayer(): number { return 3; }

  override setRoom(room: Room): void {
    super.setRoom(room);
    this.setTilesCollidable(true);
  }

  private setTilesCollidable(v: boolean): void {
    const sx = Math.floor((this.getPosition().x - this.getHalfSize().x) / 10);
    const sy = Math.floor((this.getPosition().y - this.getHalfSize().y) / 10);
    const height = Math.floor(this.getSize().y / 10);
    for (let y = sy; y < sy + height; y++)
      this.myRoom.setCollidable(sx, y, v);
  }

  override onBossWallActivate(): void {
    this.active = true;
    this.frameCounter = 0;
  }

  override onBossWallDeactivate(): void { this.active = false; }
  override onRespawn(): void { this.active = false; }

  override draw(ctx: CanvasRenderingContext2D, offsetX: number, offsetY: number, _layer: number): void {
    const x = Math.round(this.getDrawPositionX() + offsetX - this.getHalfSize().x);
    const y = Math.round(this.getDrawPositionY() + offsetY - this.getHalfSize().y);
    this.wall.drawFrame(ctx, 0, x, y, this.direction === Direction.LEFT, false);
  }

  override update(): void {
    this.frameCounter++;

    if (this.myRoom.getHero().Collides(this)) this.myRoom.getHero().kill();
    if (!this.active) return;

    if (this.direction === Direction.RIGHT) {
      if (this.frameCounter % 200 === 0) this.spawnSaw(-10);
      if ((100 + this.frameCounter) % 200 === 0) this.spawnSaw(-30);
      if ((180 + this.frameCounter) % 200 === 0) this.spawnSaw(-50);
    }

    if (this.direction === Direction.LEFT) {
      if (this.frameCounter % 350 === 0) this.spawnSaw(-10);
      if ((100 + this.frameCounter) % 350 === 0) this.spawnSaw(-30);
      if ((300 + this.frameCounter) % 350 === 0) this.spawnSaw(-50);
    }
  }

  private spawnSaw(yOffset: number): void {
    const saw = new BossSaw(this.direction);
    saw.setPosition(new ImmutableVec2(this.getPosition().x, this.getPosition().y + this.getHalfSize().y + yOffset));
    this.myRoom.addEntity(saw);
  }
}

File: wii/src/entities/boss_wall.cpp
#include "entities/boss_wall.h"
#include "entities/boss_saw.h"
#include "entities/hero.h"
#include "resource.h"
#include "media/animation.h"
#include "room.h"
#include <cmath>

BossWall::BossWall(Direction direction)
    : direction_(direction) {
    setSize(Vec2(10, 100));
    wall_ = Resource::getAnimation("data/images/wall.bmp", 1);
}

void BossWall::setRoom(Room* room) {
    Entity::setRoom(room);
    setTilesCollidable(true);
}

void BossWall::setTilesCollidable(bool v) {
    int sx = static_cast<int>(std::floor((position_.x - halfSize_.x) / 10.0f));
    int sy = static_cast<int>(std::floor((position_.y - halfSize_.y) / 10.0f));
    int height = static_cast<int>(size_.y) / 10;
    for (int y = sy; y < sy + height; y++)
        room_->setCollidable(sx, y, v);
}

void BossWall::onBossWallActivate() {
    active_ = true;
    wallFrameCounter_ = 0;
}

void BossWall::onBossWallDeactivate() {
    active_ = false;
}

void BossWall::onRespawn() {
    active_ = false;
}

void BossWall::draw(SDL_Renderer* renderer, int offsetX, int offsetY, int /*layer*/) {
    int x = static_cast<int>(std::round(getDrawPositionX() + offsetX - halfSize_.x));
    int y = static_cast<int>(std::round(getDrawPositionY() + offsetY - halfSize_.y));
    wall_->drawFrame(renderer, 0, x, y, direction_ == Direction::LEFT, false);
}

void BossWall::update() {
    wallFrameCounter_++;

    if (room_->getHero()->Collides(*this)) room_->getHero()->kill();
    if (!active_) return;

    if (direction_ == Direction::RIGHT) {
        if (wallFrameCounter_ % 200 == 0) spawnSaw(-10.0f);
        if ((100 + wallFrameCounter_) % 200 == 0) spawnSaw(-30.0f);
        if ((180 + wallFrameCounter_) % 200 == 0) spawnSaw(-50.0f);
    }

    if (direction_ == Direction::LEFT) {
        if (wallFrameCounter_ % 350 == 0) spawnSaw(-10.0f);
        if ((100 + wallFrameCounter_) % 350 == 0) spawnSaw(-30.0f);
        if ((300 + wallFrameCounter_) % 350 == 0) spawnSaw(-50.0f);
    }
}

void BossWall::spawnSaw(float yOffset) {
    BossSaw* saw = new BossSaw(direction_);
    saw->setPosition(Vec2(position_.x, position_.y + halfSize_.y + yOffset));
    room_->addEntity(saw);
}
